keps/sig-auth/20190408-node-scoped-daemonset.md

```diff
@@ -0,0 +1,239 @@
+  using the pod info, NodeAuthorizer can get the node information of the request
+  sender by checking its resource graph (require changing
+  [pkg/auth/nodeidentifier](https://github.com/kubernetes/kubernetes/tree/master/pkg/auth/nodeidentifier)
+  and NodeAuthorizer);
+* At this point, NodeAuthorizer has all the necessary information to make its
+  authorization decision. The graph below shows the decision making process.
+
+![The decision-making process of NodeAuthorizer](node-scoped-daemonset.png)
+
+### Workflow of NodeRestriction for DaemonSet Requests
+
+* A DaemonSet pod P1, from Node N1 sends a request to APIServer;
+* APIServer authenticates the request, adds the pod name, the pod UID, and the
+  node-scoped info into the Extra field of user.DefaultInfo;
+* NodeRestriction gets the pod info and the node-scoped info from the Extra
+  field of user.DefaultInfo through GetExtra; using the pod info,
+  NodeRestriction can gets the node information of P1 by checking
+  [Pod.Spec.NodeName](https://github.com/kubernetes/kubernetes/blob/d99f49d2a7caa2e86cb27541a267577771e133a7/staging/src/k8s.io/api/core/v1/types.go#L2874);
+  (require changing
+  [pkg/auth/nodeidentifier](https://github.com/kubernetes/kubernetes/tree/master/pkg/auth/nodeidentifier)
+  and NodeRestriction);
+* If the DaemonSet is not node-scoped, NodeRestriction returns immediately;
+  otherwise, the existing NodeRestriction logic executes.
+
+### Implementation Details
+
+* Add a new interface for getting the node information from pod name and UID
+
+```golang
+type NodeInfo interface {
+    // Given podName and podUID, figure out the node name of a pod.
+    NodeName(podName, podUID string) string
+}
+
+```
+
+Both NodeAuthorizer and NodeRestriction should implement the NodeInfo interface.
+NodeAuthroizer gets the node name of a pod by checking its resource graph; while
+NodeRestriction gets the node name of a pod by inspecting the pod spec.
+
+* Change the
+  [NodeIdentifier](https://github.com/kubernetes/kubernetes/blob/b274fb35708c8b33445fc2f5f3b006921582b1e0/pkg/auth/nodeidentifier/interfaces.go#L24)
+  interface
+
+```golang
+// given user.Info and NodeInfo, NodeIdentity decides whether the request sender is
+// a Kubelet or a node-scoped DaemonSet pod. If the request sender is a Kubelet,
+// isKubelet will be set to true, and nodeName will be set to the node name; if the
+// request sender is a node-scoped DaemsonSet pod, isNodeScopedDaemonSet will be
+// set to true, and nodeName will be set to the node name.
+NodeIdentity(user.Info, NodeInfo) (nodeName string, isKubelet bool, isNodeScopedDaemonSet bool)
+```
+
+## Design Details
+
+### Test Plan
+
+The existing unit and integration tests of NodeAuthorizer and NodeRestriction
+will be extended to cover requests from node-scoped Daemonsets.
+
+### Graduation Criteria
+
+v1.15:
+
+* implement the proposal, and extend the existing unit and integration tests.
```

Comment:
implement the proposal in alpha, right? This should have a dedicated feature gate, and will probably need an API review.

Reply:
Agree with dedicated feature gate

Reply:
Thanks for pointing out. Fixed it!